[MIG][10.0] account_cutoff_accrual_picking

TDu opened the pull request:
Hello,
Started migrating account_cutoff_accrual_picking, some schema changes impact this module.
It's a work in progress.
Still looking for a way to calculate them for a previous date correctly.
Cheers.

TDu committed:
[MIG][10.0] account_cutoff_accrual_picking

Set sale/purchase order in db, fix default account for new accrued
Fix the processing of taxes and the journal entries for them
Fix exception deprecated and with_context on currency change

guewen commented on account_cutoff_accrual_picking/models/account_cutoff.py:
```diff
@@ -0,0 +1,234 @@
+                tax_accrual_amount = self.env['res.currency'].with_context(
+                    date=self.cutoff_date)._compute(currency,
+                                                    company_currency_id,
+                                                    tax_line['amount'])
```
IMO it is cleaner to use the public `compute` method:
```
    @api.multi
    def compute(self, from_amount, to_currency, round=True):
        """ Convert `from_amount` from currency `self` to `to_currency`. """
```
Which gives

```python
currency_at_date = currency.with_context(date=self.cutoff_date)
tax_accrual_amount = currency_at_date.compute(company_currency, tax_line['amount'])

TDu replied:
Much better, thank you, it is changed

guewen commented on account_cutoff_accrual_picking/models/account_cutoff.py:
```diff
@@ -0,0 +1,234 @@
+        if company_currency_id != currency_id:
+            amount_company_currency = self.env['res.currency'].with_context(
+                date=self.cutoff_date)._compute(
+                    currency, company_currency_id, amount)
```
Same as above, regarding usage of `compute` vs `_compute`

TDu replied:
Changed

guewen commented on account_cutoff_accrual_picking/models/account_cutoff.py:
```diff
@@ -0,0 +1,234 @@
+                account_id = line.product_id.product_tmpl_id.categ_id.\
+                    property_account_expense_categ_id.id
+            if not account_id:
+                raise exceptions.Warning(
```
`Warning` is deprecated, use `exceptions.UserError`

TDu replied:
Ok, changed

guewen commented on account_cutoff_accrual_picking/models/account_cutoff.py:
```diff
@@ -0,0 +1,234 @@
+                raise exceptions.Warning(
+                    _("Error: Missing expense account on product '%s' or on "
+                        "related product category.") % (line.product_id.name))
+            analytic_account_id = line.account_analytic_id.id or False
```
`or False` can be removed: if there is no analytic account, it will return an empty record which will be evaluated as `False` too

TDu replied:
Thanks, removed

guewen commented on account_cutoff_accrual_picking/models/account_cutoff.py:
```diff
@@ -0,0 +1,234 @@
+                account_id = line.product_id.product_tmpl_id.categ_id.\
+                    property_account_income_categ_id.id
+            if not account_id:
+                raise exceptions.Warning(
```
`UserError`

TDu replied:
Changed

guewen commented on account_cutoff_accrual_picking/models/account_cutoff.py:
```diff
@@ -0,0 +1,234 @@
+            tax_read = self.env['account.tax'].browse(tax_line['id'])
+            tax_accrual_account_id = tax_read[tax_account_field_name]
+            if not tax_accrual_account_id:
+                raise exceptions.Warning(
```
`UserError`

TDu replied:
Changed

guewen commented on account_cutoff_accrual_picking/README.rst:
```diff
@@ -0,0 +1,17 @@
+## Usage
+
+
+
```
Need to use this template: https://raw.githubusercontent.com/OCA/maintainer-tools/master/template/module/README.rst

gurneyalex commented:
@TDu  any news on this ?

TDu commented:
Nope I have not work on it for a while.

jbaudoux committed:
account_cutoff_accrual_picking: take care of fiscal positions for account and tax mapping

jbaudoux committed:
fix currency (comparing apple and orange) ; remove obsolete stock_move_id in cutoff line ; add product_id in cutoff line

jbaudoux commented:
Hello Thierry,
I reviewed and tested your PR. Here is a PR that makes some fixes. Can you merge it in this branch?
https://github.com/TDu/account-closing/pull/1

TDu commented:
@jbaudoux Done

jbaudoux committed:
add cutoff cron jobs

jbaudoux commented:
@TDu Hello Thierry,
Here is a second PR that workaround the fact that we cannot generate accrual at date. I have added cron jobs that ensures accrual is performed on time at the end of period. See https://github.com/TDu/account-closing/pull/3

TDu commented:
@jbaudoux Hello,
From what I can remember calculating the accrual for another time than now will not be accurate because the data it is calculated on will/may have changed.
So saving the a date different that the time of calculation is a little bit misleading, no ?

jbaudoux commented:
@Tdu. That's correct. The accrual cannot be run at date as we don't have an history of qty to invoice on the sale order line. As a consequence, it must be run at the end of the last day of the month, this is why I'm adding a cron job doing this. The accrual must be dated on the last day of the month and reversed on the first day of next month. I configured the cron job to run at the first day of the month at 00:00 AM. (To be adjusted by user depending on timezone). That's the easiest to express that it must run each month. I included some time safety code by computing the right last day no matter the job run on the last day of the month or on the first day of next month.

TDu commented:
Ok thanks for the explanation.

oleksandrpaziuk commented on account_cutoff_accrual_picking/models/account_cutoff.py:
```diff
@@ -0,0 +1,268 @@
+            if company_currency != currency:
+                currency_at_date = currency.with_context(date=self.cutoff_date)
+                tax_accrual_amount = currency_at_date.compute(
+                    tax_line['amount'], company_currency.id)
```
change `company_currency.id` to `company_currency`

jbaudoux replied:
Where is that compute method on currency defined? Can't find it :(

TDu replied:
It is here https://github.com/odoo/odoo/blob/3bd2bb7aa1e3448cd946cbe86533051e139b6258/odoo/addons/base/res/res_currency.py#L149
His fix looks good to me, I asked him to propose the change

oleksandrpaziuk replied:
`odoo/addons/base/res/res_currency.py`

oleksandrpaziuk replied:
please review pr https://github.com/TDu/account-closing/pull/4

TDu replied:
Thanks merged.

Commit:
fix account_cutoff_accrual_picking currency_at_date compute

leemannd reviewed (approved):
Approved once travis is green @jbaudoux Do you know why it went red with your commit?

fclementic2c commented:
@jbaudoux Any chance that we get this module merged soon ? thxs

TDu committed:
Fix test isolate account_multicurrency_revaluation

The tests of account_multicurrency_revaluation do not like the demo
data from account module.

fclementic2c commented:
![image](https://user-images.githubusercontent.com/5735158/37767002-fbaf426e-2dc8-11e8-9fc7-f2e502842918.png)

jbaudoux commented:
See new PR https://github.com/OCA/account-closing/pull/73
This one can be closed.
